# HUFFINGTON POST DATASET CATEGORY LABEL ML MODEL

In [59]:
import json
import numpy as np
import os

Use json module to load the file

In [60]:
data = []
with open(r'F:\notes\News_Category_Dataset.json\News_Category_Dataset.json') as f:
    for line in f:
        data.append(json.loads(line))
        

In [61]:
category_list = []
headline_list = []
description_list = []
author_list = []
label_list = []
all_text = []

Create separate lists for each attribute and fill them with all the data for subsequent attribute.

In [62]:
for f in data:
    category_list.append(f.get('category'))
    headline_list.append(f.get('headline'))
    description_list.append(f.get('short_description'))

    

In [63]:
print("category:",category_list[1],
"headline:",headline_list[1],
"description:",description_list[1])

category: ENTERTAINMENT headline: Will Smith Joins Diplo And Nicky Jam For The 2018 World Cup's Official Song description: Of course it has a song.


In [64]:
labels = {'POLITICS': 1,
'ENTERTAINMENT': 2,
'HEALTHY LIVING': 3,
'QUEER VOICES': 4,
'BUSINESS': 5,
'SPORTS': 6,
'COMEDY': 7,
'PARENTS': 8,
'BLACK VOICES': 9,
'THE WORLDPOST': 10,
'WOMEN': 11,
'CRIME': 12,
'MEDIA': 13,
'WEIRD NEWS': 14,
'GREEN': 15,
'IMPACT': 16,
'WORLDPOST': 17,
'RELIGION': 18,
'STYLE': 19,
'WORLD NEWS': 20,
'TRAVEL': 21,
'TASTE': 22,
'ARTS': 23,
'FIFTY': 24,
'GOOD NEWS': 25,
'SCIENCE': 26,
'ARTS & CULTURE': 27,
'TECH': 28,
'COLLEGE': 29,
'LATINO VOICES': 30,
'EDUCATION': 31}


Right now, Labels i.e. category  are in simple integer 
,we need to convert it into one hot encode so the model can understand

In [65]:
def one_hot_encoder(dictionary,number):
    
    for f in dictionary:
        out = np.zeros(number)
        n = dictionary.get(f)
        out[n-1] = 1
        dictionary[f] = out
    return dictionary

In [66]:
out=one_hot_encoder(labels,31)
out

{'POLITICS': array([1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'ENTERTAINMENT': array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'HEALTHY LIVING': array([0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'QUEER VOICES': array([0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'BUSINESS': array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'SPORTS': array([0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 'COMEDY': array([0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0

You can use this encoder function that i made or you can use tflearn and keras built-in functions to_categorical

In [67]:
for f in category_list:
    label_list.append(out.get(f))

Now lets convert our category list into one hot encoded list and store in label_ist

In [68]:
label_list[:5]

[array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])]

In [69]:
for i ,j in zip(description_list, headline_list):
    all_text.append(i+j)
    

Right now , description and headline are separate.Lets concatenate them so that model can train both on headlines as well as description

In [70]:
import tensorflow as tf
from tensorflow.python.keras.models import Sequential
from tensorflow.python.keras.layers import Dense, GRU, Embedding,LSTM
from tensorflow.python.keras.optimizers import Adam
from tensorflow.python.keras.preprocessing.text import Tokenizer
from tensorflow.python.keras.preprocessing.sequence import pad_sequences

In [71]:
tokenizer = Tokenizer(num_words=None)


In [72]:
tokenizer.fit_on_texts(all_text)

In [73]:
tokenizer.word_index

{'the': 1,
 'to': 2,
 'a': 3,
 'of': 4,
 'and': 5,
 'in': 6,
 'is': 7,
 'for': 8,
 'on': 9,
 'that': 10,
 'with': 11,
 'it': 12,
 'you': 13,
 'are': 14,
 'i': 15,
 'this': 16,
 'be': 17,
 'as': 18,
 'at': 19,
 'trump': 20,
 'from': 21,
 'have': 22,
 'not': 23,
 'we': 24,
 'about': 25,
 'was': 26,
 'new': 27,
 'his': 28,
 'but': 29,
 'your': 30,
 'has': 31,
 'an': 32,
 'what': 33,
 'will': 34,
 'by': 35,
 'he': 36,
 'my': 37,
 'one': 38,
 'all': 39,
 'how': 40,
 'more': 41,
 'who': 42,
 'can': 43,
 'their': 44,
 'up': 45,
 'people': 46,
 'after': 47,
 'out': 48,
 'our': 49,
 'her': 50,
 'just': 51,
 'they': 52,
 'when': 53,
 'or': 54,
 'donald': 55,
 "it's": 56,
 'like': 57,
 'if': 58,
 'time': 59,
 'over': 60,
 'so': 61,
 'do': 62,
 'no': 63,
 'says': 64,
 'than': 65,
 'said': 66,
 'why': 67,
 'year': 68,
 'first': 69,
 'been': 70,
 's': 71,
 'day': 72,
 'u': 73,
 'president': 74,
 'get': 75,
 'world': 76,
 'women': 77,
 'life': 78,
 'some': 79,
 'most': 80,
 'now': 81,
 'make': 82,
 '

keras.tokenizer allows us to turn each unique word in our all_text list into a unique integer.since ML model deals with integers not words.

In [74]:
from sklearn.model_selection import train_test_split
all_text = np.array(all_text)
label_list = np.array(label_list)
X_train,X_test,y_train,y_test = train_test_split(all_text,label_list,test_size=0.3)

Split the dataset and labels into training and testing dataset.

In [75]:
x_train_tokens = tokenizer.texts_to_sequences(X_train)

Now convert the lists in all_text to sequences of integers of their corresponding words

In [77]:
print(X_train[15000],
x_train_tokens[15000])

LISTEN: This Is What The Sun 'Sounds' Like [1324, 16, 7, 33, 1, 2553, 96001, 57]


In [78]:
x_test_tokens = tokenizer.texts_to_sequences(X_test)

Now the lists are of different lengths .To feed in our model , we have to make all our lists of same size

In [79]:
num_tokens = [len(tokens) for tokens in x_train_tokens + x_test_tokens]
num_tokens = np.array(num_tokens)

In [80]:
np.mean(num_tokens)

25.83365736184784

In [81]:
np.max(num_tokens)

248

In [82]:
max_tokens = np.mean(num_tokens) + 2 * np.std(num_tokens)
max_tokens = int(max_tokens)
max_tokens

54

In [83]:
np.sum(num_tokens < max_tokens) / len(num_tokens)

0.9446111257790686

We now pad small length lists and truncate long lists , so that all lists are of 54 length.

In [84]:
from tflearn.data_utils import pad_sequences,to_categorical
x_train_pad = pad_sequences(x_train_tokens,maxlen=max_tokens)
x_test_pad = pad_sequences(x_test_tokens,maxlen=max_tokens)

In [85]:
x_train_pad.shape

(87492, 54)

In [86]:
x_test_pad.shape

(37497, 54)

In [87]:
idx = tokenizer.word_index
inverse_map = dict(zip(idx.values(), idx.keys()))

In [88]:
def tokens_to_string(tokens):
    # Map from tokens back to words.
    words = [inverse_map[token] for token in tokens if token != 0]
    
    # Concatenate all words.
    text = " ".join(words)

    return text

In [89]:
len(tokenizer.word_index) #total no. of words in our vocab

99572

In [90]:
tokens_to_string(x_train_tokens[1])

'sitting is the new smoking and a hazard to your career health'

In [91]:
x_train_tokens[1]

[1931, 7, 1, 27, 3259, 5, 3, 12470, 2, 30, 690, 122]

In [92]:
#pad = 'pre'
#x_train_pad = pad_sequences(x_train_tokens, maxlen=max_tokens,
#                           padding=pad, truncating=pad)
#x_test_pad = pad_sequences(x_test_tokens, maxlen=max_tokens,
#                           padding=pad, truncating=pad)
#model = Sequential()
#embedding_size = 128
#model.add(Embedding(input_dim=len(tokenizer.word_index),
#                    output_dim=embedding_size,
#                    input_length=max_tokens,
#                    name='layer_embedding'))
#optimizer = Adam(lr=1e-3)
#model.add(LSTM(128,dropout=0.2))
#model.add(Dense(31, activation='softmax'))
#model.compile(loss='categorical_crossentropy',
#              optimizer='adam',
#              metrics=['accuracy'])
#model.summary()
#model.fit(x_train_pad, y_train,
#          epochs=3, batch_size=64)
#score = model.evaluate(x_test_pad, y_test, batch_size=64)
#print("Accuracy: {0:.2%}".format(score[1]))

Above is keras model and below is tflearn model u can choose either , better accuracy with tflearn , i think so

In [93]:
import tflearn
tf.reset_default_graph()
RNN = tflearn.input_data([None,54])
RNN = tflearn.embedding(RNN,input_dim=99573,output_dim=128)
RNN = tflearn.lstm(RNN,128,dropout=0.8)
RNN = tflearn.fully_connected(RNN,31,activation='softmax')
RNN = tflearn.regression(RNN,optimizer='adam',learning_rate=0.001,loss='categorical_crossentropy')


In [94]:

model = tflearn.DNN(RNN,tensorboard_verbose=0)
model.fit(x_train_pad,y_train,n_epoch =5,validation_set=(x_test_pad,y_test),show_metric=True,batch_size=32,run_id='huff')


Training Step: 13674  | total loss: 0.39250 | time: 591.191s
| Adam | epoch: 005 | loss: 0.39250 - acc: 0.8829 -- iter: 87488/87492
Training Step: 13675  | total loss: 0.42521 | time: 613.890s
| Adam | epoch: 005 | loss: 0.42521 - acc: 0.8790 | val_loss: 1.85761 - val_acc: 0.5730 -- iter: 87492/87492
--


change tensorboard_verbose = 3 if you want to visualise using tensorboard and dont increase n_epochs >5 ,model may overfit.The input_dim in 
tflearn is len(tokenizer.word_index) + 1 compared to keras which is len(tokenizer.word_index) i.e 99572.



In [95]:
model.save('quickest.model')

INFO:tensorflow:C:\python36\huff\quickest.model is not in all_model_checkpoint_paths. Manually adding it.


In [113]:
labelo = {'POLITICS': 1,
'ENTERTAINMENT': 2,
'HEALTHY LIVING': 3,
'QUEER VOICES': 4,
'BUSINESS': 5,
'SPORTS': 6,
'COMEDY': 7,
'PARENTS': 8,
'BLACK VOICES': 9,
'THE WORLDPOST': 10,
'WOMEN': 11,
'CRIME': 12,
'MEDIA': 13,
'WEIRD NEWS': 14,
'GREEN': 15,
'IMPACT': 16,
'WORLDPOST': 17,
'RELIGION': 18,
'STYLE': 19,
'WORLD NEWS': 20,
'TRAVEL': 21,
'TASTE': 22,
'ARTS': 23,
'FIFTY': 24,
'GOOD NEWS': 25,
'SCIENCE': 26,
'ARTS & CULTURE': 27,
'TECH': 28,
'COLLEGE': 29,
'LATINO VOICES': 30,
'EDUCATION': 31}


In [116]:
def predictcategory(user_input):
    
    sentence  = tokenizer.texts_to_sequences(user_input)
    sentence = pad_sequences(sentence,maxlen=max_tokens)
    result = model.predict(sentence)
    result = np.argmax(result)
    result += 1
    for f in labelo:
        if labelo.get(f) == result:
            answer = f
    return answer
    


In [121]:
yes = predictcategory(['books for graduation schools will provide'])

In [122]:
yes

'EDUCATION'

1